components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,281 @@
+	}
+	if err := at.IntrinsicParams.CheckValid(); err != nil {
+		return err
+	}
+	if err := at.DistortionParams.CheckValid(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	u                       *url.URL
+	client                  *gortsplib.Client
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	latestFrame             atomic.Value
+	logger                  golog.Logger
+}
+
+// Close closes the camera.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	var clientTerminated liberrors.ErrClientTerminated
+	rc.cancelFunc()
+	rc.activeBackgroundWorkers.Wait()
+	if err := rc.client.Close(); err != nil && !errors.Is(err, clientTerminated) {
+		return err
+	}
+	return nil
+}
+
+// clientReconnectBackgroundWorker checks every 5 sec to see if the client is connected to the server, and reconnects if not.
+func (rc *rtspCamera) clientReconnectBackgroundWorker() {
+	rc.activeBackgroundWorkers.Add(1)
+	ticker := time.NewTicker(5 * time.Second)
+	goutils.ManagedGo(func() {
+		for {
+			select {
+			case <-rc.cancelCtx.Done():
+				return
+			case <-ticker.C:
+				// check if the client is still connected with an OPTIONS request
+				res, err := rc.client.Options(rc.u)
+				if err != nil && (errors.Is(err, liberrors.ErrClientTerminated{}) ||
+					strings.Contains(err.Error(), "EOF") ||
+					strings.Contains(err.Error(), "connection refused") ||
+					strings.Contains(err.Error(), "broken pipe")) {
+					rc.logger.Warnf("The rtsp client for %s has error %s", rc.u, err)
+					if err = restartClient(rc); err != nil {
+						rc.logger.Warnf("cannot reconnect to rtsp server: %s", err)
+					} else {
+						rc.logger.Infof("reconnected to rtsp server %s", rc.u)
+					}
+				} else if res != nil && res.StatusCode != base.StatusOK {
+					rc.logger.Warnf("The rtsp server responded with %s, trying to reconnect", res.StatusCode)
+					if err = restartClient(rc); err != nil {
+						rc.logger.Warnf("cannot reconnect to rtsp server: %s", err)
+					} else {
+						rc.logger.Infof("reconnected to rtsp server %s", rc.u)
+					}
+				}
+			}
+		}
+	}, rc.activeBackgroundWorkers.Done)
+}
+
+// NewRTSPCamera creates a camera client for an RTSP given given the server URL.
+// Right now, only supports servers that have MJPEG video tracks.
+func NewRTSPCamera(ctx context.Context, attrs *Attrs, logger golog.Logger) (camera.Camera, error) {
+	// parse URL
+	u, err := url.Parse(attrs.Address)
+	if err != nil {
+		return nil, err
+	}
+	client := &gortsplib.Client{}
+	// connect to the server - be sure to close it if setup fails.
+	var clientSuccessful bool
+	var clientErr error
+	err = client.Start(u.Scheme, u.Host)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if !clientSuccessful {
+			clientErr = client.Close()
```

Comment:
assigned clientErr here does nothing I believe. You'd need to assign a function return name to error or just log

Reply:
changed this logic such that if there was a useful error on client.Close, it would be merged with the error within this deferred function, and named the return error as well